strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+            raise ValueError("Cannot apply loss channel, mode does not exist")
+        X = np.sqrt(T) * np.identity(2)
+        Y = self.hbar * (1 - T) * (2 * nbar + 1) * np.identity(2) / 2
+        X2, Y2 = self.expandXY([k], X, Y)
+        self.apply_channel(X2, Y2)
+
+    def init_thermal(self, nbar, mode):
+        r"""Initializes a state of mode in a thermal state with the given population.
+
+        Args:
+            nbar (float): mean photon number of the thermal state
+            mode (int): mode that get initialized
+        """
+        self.thermal_loss(0.0, nbar, mode)
+
+    def is_vacuum(self, tol=0.0):
+        r"""Checks if the state is vacuum by calculating its fidelity with vacuum.
+
+        Args:
+            tol (float): the closeness tolerance to fidelity of 1
+        """
+        fid = self.fidelity_vacuum()
+        return np.abs(fid - 1) <= tol
+
+    def measure_dyne(self, covmat, indices, shots=1):
+        r"""Performs general-dyne measurements on a set of modes.
+
+        For more information see Quantum Continuous Variables: A Primer of Theoretical Methods
+        by Alessio Serafini page 129.
+
+        Args:
+            covmat (array): covariance matrix of the generaldyne measurement
+            indices (list): modes to be measured
+            shots (int): how many measurements are performed
+
+        Returns:
+            array: measurement outcome corresponding to a point in phase space
+        """
+        # pylint: disable=too-many-branches
+        if covmat.shape != (2 * len(indices), 2 * len(indices)):
+            raise ValueError("Covariance matrix size does not match indices provided")
+
+        if np.linalg.det(covmat) < (self.hbar / 2) ** (2 * len(indices)):
+            raise ValueError("Measurement covariance matrix is unphysical.")
+
+        if self.covs.imag.any():
+            raise NotImplementedError("Covariance matrices must be real")
+
+        for i in indices:
+            if self.active[i] is None:
+                raise ValueError("Cannot apply measurement, mode does not exist")
+
+        expind = np.concatenate((2 * np.array(indices), 2 * np.array(indices) + 1))
+        vals = np.zeros((shots, 2 * len(indices)))
+        imag_means_ind = np.where(self.means[:, expind].imag.any(axis=1))[0]
+        nonneg_weights_ind = np.where(np.angle(self.weights) != np.pi)[0]
+        ub_ind = np.union1d(imag_means_ind, nonneg_weights_ind)
+        ub_weights = np.abs(np.array(self.weights))
+        if imag_means_ind.size > 0:
+            ub_weights[imag_means_ind] *= np.exp(
+                0.5
+                * np.einsum(
+                    "...j,...jk,...k",
+                    (self.means[imag_means_ind, :][:, expind].imag),
+                    np.linalg.inv(
+                        self.covs[imag_means_ind, :, :][:, expind, :][:, :, expind].real + covmat
+                    ),
+                    (self.means[imag_means_ind, :][:, expind].imag),
```

Comment:
Would suggest creating objects with descriptive names so that this operation is more readable (e.g., an object for the result of `einsum`, etc.).

This would also apply to upcoming occurrences of more complex multi-line function calls (`einsum`, `np.sum`)

Reply:
I have added extensive commenting and clearer variable names now